### DISCUSSION TITLE
Call to unity from OnClick not working.

### DISCUSSION THREAD

**Discussion by Attackonteentitans** [jeffreylanters/react-unity-webgl#271] [Help Wanted!]
WHen my program executes this code to interact from React to Unity it does nothing because the this.unityInstance is null.  Why would this be since the unity application is running.  My application runs, and is now sending data to my react app, but when I try to call a method back again it fails due to this problem.  I have tried everything, How do I fix it?

UnityContext.prototype.send = function (gameObjectName, methodName, parameter) {
        if (this.unityInstance !== null) {
            if (parameter === undefined) {
                this.unityInstance.SendMessage(gameObjectName, methodName);
            }

            else {
                this.unityInstance.SendMessage(gameObjectName, methodName, parameter);
            }
        }
    };

Seems to work ifcI immediately call a UNITY method from the same method I receive a message from unity in but not when I make a call from onclick:

This works to call back..
useState(function () {
      unityContext.on("NextPlayer", function (gridString) {
        setGridString(gridString);
        if (playerNbr === 1)
        {
          playerNbr = 2;
          unityContext.send("Player2Spawner", "nextPlayer", gridString);
        }
        else {
          playerNbr = 1;
          unityContext.send("Player1Spawner", "nextPlayer", gridString);
        }

        });
      }, []);

**Comment by jeffreylanters** [comment]
Hi. Could you send a little more code? I'm interested in the part where you declare the Unity Context. Depending on your implementation, you might break its reference to the Unity Application.

**Reply by Attackonteentitans** [reply]
const unityContext = new UnityContext({
  loaderUrl: "Anti-Matter Tetris WEBGL/build/Anti-Matter Tetris.loader.js",
  dataUrl: "Anti-Matter Tetris WEBGL/build/Anti-Matter Tetris.data",
  frameworkUrl: "Anti-Matter Tetris WEBGL/build/Anti-Matter Tetris.framework.js",
  codeUrl: "Anti-Matter Tetris WEBGL/build/Anti-Matter Tetris.wasm",
  });

    const [isGameOver, setIsGameOver] = useState(false);
    const [gridString, setGridString] = useState("");
    const [didError, setDidError] = useState(false);
    const [errorMessage, setErrorMessage] = useState("");
    var playerNbr = 1;


    useState(function () {
      unityContext.on("GameOver", function () {
        setIsGameOver(true);
        });
      }, []);

    useState(function () {
      unityContext.on("NextPlayer", function (gridString) {
        setGridString(gridString);
        if (playerNbr === 1)
        {
          playerNbr = 2;
          unityContext.send("Player2Spawner", "nextPlayer", gridString);
        }
        else {
          playerNbr = 1;
          unityContext.send("Player1Spawner", "nextPlayer", gridString);
        }

        });
      }, []);

    function Player2NextTurn()
        {
        console.log("Player 2 clicked");
        unityContext.send("Player2Spawner", "nextPlayer", gridString);
        }

    function Player1NextTurn()
        {
        console.log("Player 1 clicked");
        unityContext.send("Player1Spawner", "nextPlayer", gridString);
        }


And then this is the code that does not work:

  <Unity unityContext={unityContext}
              style={{
                  height: "720px",
                  width: "600px",
                  border: "2px solid black",
                  background: "grey",
                  }}
          />

      <button onClick={Player1NextTurn}>Player 1 Next Turn</button>
      <button onClick={Player2NextTurn}>Player 2 Next Turn</button>

The player 1 clicked and the player 2 clicked is logged to the console but when I debugged into the code below where the .send is executed, it just returns and does nothing as the this.unityInstance is null.

UnityContext.prototype.send = function (gameObjectName, methodName, parameter) {
if (this.unityInstance !== null) {
if (parameter === undefined) {
this.unityInstance.SendMessage(gameObjectName, methodName);
}

**Reply by Attackonteentitans** [reply]
wheras the same code executes as expected here:

useState(function () {
  unityContext.on("NextPlayer", function (gridString) {
    setGridString(gridString);
    if (playerNbr === 1)
    {
      playerNbr = 2;
      unityContext.send("Player2Spawner", "nextPlayer", gridString);
    }
    else {
      playerNbr = 1;
      unityContext.send("Player1Spawner", "nextPlayer", gridString);
    }

    });
  }, []);

**Reply by jeffreylanters** [reply]
Could you try moving your UnityContext creation _outside_ of your render method? Like this:

```js
import React from "react";
import Unity, { UnityContext } from "react-unity-webgl";

const unityContext = new UnityContext({
  loaderUrl: "build/myunityapp.loader.js",
  dataUrl: "build/myunityapp.data",
  frameworkUrl: "build/myunityapp.framework.js",
  codeUrl: "build/myunityapp.wasm",
});

function App() {
  return <Unity unityContext={unityContext} />;
}
```

**Reply by Attackonteentitans** [reply]
OK. I will try that..

**Reply by Attackonteentitans** [reply]
thank you. That fixed the issue..

**Reply by jeffreylanters** [reply]
Awesome! Could you mark this thread as the solution? Happy coding!
